File: scripts/python/999_logging_config.py
import logging
import os
import sys
from datetime import datetime

# Color codes for different script types
SCRIPT_COLORS = {
    'python': '\033[36m',      # Cyan
    'bash': '\033[32m',        # Green
    'javascript': '\033[33m',  # Yellow
    'sql': '\033[35m',         # Magenta
    'system': '\033[90m',      # Dark Gray
    'reset': '\033[0m'         # Reset
}

# Log level colors
LEVEL_COLORS = {
    'DEBUG': '\033[37m',       # White
    'INFO': '\033[32m',        # Green
    'WARNING': '\033[33m',     # Yellow
    'ERROR': '\033[31m',       # Red
    'CRITICAL': '\033[41m',    # Red background
    'reset': '\033[0m'
}

def get_script_type(script_name):
    """Determine script type from filename extension or name pattern."""
    if script_name.endswith('.py') or 'python' in script_name.lower():
        return 'python'
    elif script_name.endswith('.sh') or 'bash' in script_name.lower():
        return 'bash'
    elif script_name.endswith('.js') or 'node' in script_name.lower():
        return 'javascript'
    elif script_name.endswith('.sql'):
        return 'sql'
    else:
        return 'system'

def setup_logging(script_name, log_dir=None, level=logging.INFO):
    """
    Sets up a standardized logger for a script.

    Args:
        script_name (str): The name of the script (e.g., os.path.basename(__file__).replace('.py', '')).
        log_dir (str, optional): The directory where log files will be stored.
                                 Defaults to '~/log'.
        level (int, optional): The logging level (e.g., logging.INFO, logging.DEBUG).
                               Defaults to logging.INFO.
    Returns:
        logging.Logger: The configured logger instance.
    """
    if log_dir is None:
        log_dir = os.path.expanduser('~/log')

    os.makedirs(log_dir, exist_ok=True)

    # Remove all existing handlers to prevent duplicate logs
    for handler in logging.root.handlers[:]:
        logging.root.removeHandler(handler)
    for handler in logging.getLogger().handlers[:]:
        logging.getLogger().removeHandler(handler)

    logger = logging.getLogger(script_name)
    logger.setLevel(level)

    # File handler
    now = datetime.now()
    formatted_time = now.strftime('%Y-%m-%d_%H-%M-%S')
    file_handler = logging.FileHandler(os.path.join(log_dir, f"{script_name}_log_{formatted_time}.log"))
    file_handler.setLevel(level)
    file_formatter = logging.Formatter('%(asctime)s - %(levelname)s - %(message)s')
    file_handler.setFormatter(file_formatter)
    logger.addHandler(file_handler)

    # Console handler with colors
    console_handler = logging.StreamHandler(sys.stdout)
    console_handler.setLevel(logging.INFO) # Console always INFO or higher
    
    # Custom formatter with script type and colors
    script_type = get_script_type(script_name)
    script_color = SCRIPT_COLORS.get(script_type, SCRIPT_COLORS['system'])
    reset_color = SCRIPT_COLORS['reset']
    
    class ColoredFormatter(logging.Formatter):
        def format(self, record):
            # Add script type to record
            record.script_type = script_type.upper()
            record.script_name = script_name
            record.pid = os.getpid()
            
            # Color the level
            level_color = LEVEL_COLORS.get(record.levelname, LEVEL_COLORS['reset'])
            colored_level = f"{level_color}{record.levelname}{LEVEL_COLORS['reset']}"
            
            # Color the script info
            colored_script = f"{script_color}🐍{record.script_type}:{record.script_name}{reset_color}"
            
            # Create formatted message
            formatted = f"[{record.asctime}] [{colored_level}] [{colored_script}] [PID:{record.pid}] - {record.getMessage()}"
            return formatted
    
    console_formatter = ColoredFormatter('%(asctime)s')
    console_handler.setFormatter(console_formatter)
    logger.addHandler(console_handler)

    return logger

File: scripts/nodejs/999_logging_config.js
#!/usr/bin/env node
/**
 * Unified logging configuration for Node.js scripts
 * Matches the Python and Bash logging formats with consistent colors
 */

const fs = require('fs');
const path = require('path');

// Color codes for different script types and levels
const SCRIPT_COLORS = {
    javascript: '\x1b[33m',  // Yellow
    python: '\x1b[36m',      // Cyan
    bash: '\x1b[32m',        // Green
    sql: '\x1b[35m',         // Magenta
    system: '\x1b[90m',      // Dark Gray
    reset: '\x1b[0m'         // Reset
};

const LEVEL_COLORS = {
    DEBUG: '\x1b[37m',       // White
    INFO: '\x1b[32m',        // Green
    WARN: '\x1b[33m',        // Yellow
    ERROR: '\x1b[31m',       // Red
    CRITICAL: '\x1b[41m',    // Red background
    reset: '\x1b[0m'
};

const LOG_LEVELS = {
    DEBUG: 0,
    INFO: 1,
    WARN: 2,
    ERROR: 3,
    CRITICAL: 4
};

class TrilliumLogger {
    constructor(scriptName, options = {}) {
        this.scriptName = scriptName || path.basename(process.argv[1], '.js');
        this.logDir = options.logDir || path.join(process.env.HOME, 'log');
        this.logLevel = options.logLevel || 'INFO';
        this.scriptType = 'JAVASCRIPT';
        this.pid = process.pid;
        
        // Ensure log directory exists
        if (!fs.existsSync(this.logDir)) {
            fs.mkdirSync(this.logDir, { recursive: true });
        }
        
        // Setup log file
        const timestamp = new Date().toISOString().replace(/[:.]/g, '-').slice(0, 19);
        this.logFile = path.join(this.logDir, `${this.scriptName}_log_${timestamp}.log`);
    }
    
    shouldLog(level) {
        const currentPriority = LOG_LEVELS[this.logLevel] || 1;
        const messagePriority = LOG_LEVELS[level] || 1;
        return messagePriority >= currentPriority;
    }
    
    formatTimestamp() {
        return new Date().toISOString().replace('T', ' ').slice(0, 19);
    }
    
    formatLogEntry(level, message, caller = '') {
        const timestamp = this.formatTimestamp();
        const callerInfo = caller ? ` [${caller}]` : '';
        
        // Plain format for file logging
        const plainEntry = `[${timestamp}] [${level}] [${this.scriptType}:${this.scriptName}] [PID:${this.pid}]${callerInfo} - ${message}`;
        
        // Colored format for console
        const levelColor = LEVEL_COLORS[level] || LEVEL_COLORS.reset;
        const scriptColor = SCRIPT_COLORS.javascript;
        const resetColor = SCRIPT_COLORS.reset;
        
        const coloredEntry = `[${timestamp}] [${levelColor}${level}${LEVEL_COLORS.reset}] [${scriptColor}🟨${this.scriptType}:${this.scriptName}${resetColor}] [PID:${this.pid}]${callerInfo} - ${message}`;
        
        return { plainEntry, coloredEntry };
    }
    
    log(level, message, caller = '') {
        if (!this.shouldLog(level)) {
            return;
        }
        
        const { plainEntry, coloredEntry } = this.formatLogEntry(level, message, caller);
        
        // Console output with colors (to stderr)
        if (process.stdout.isTTY) {
            process.stderr.write(coloredEntry + '\n');
        } else {
            process.stderr.write(plainEntry + '\n');
        }
        
        // File output without colors
        try {
            fs.appendFileSync(this.logFile, plainEntry + '\n');
        } catch (err) {
            // Fallback - at least show the message
            process.stderr.write(`[LOG ERROR] Could not write to log file: ${err.message}\n`);
        }
    }
    
    debug(message, caller) {
        this.log('DEBUG', message, caller);
    }
    
    info(message, caller) {
        this.log('INFO', message, caller);
    }
    
    warn(message, caller) {
        this.log('WARN', message, caller);
    }
    
    error(message, caller) {
        this.log('ERROR', message, caller);
    }
    
    critical(message, caller) {
        this.log('CRITICAL', message, caller);
    }
    
    // Execution time logging
    logExecutionTime(description, startTime, level = 'INFO') {
        const endTime = Date.now();
        const duration = ((endTime - startTime) / 1000).toFixed(2);
        this.log(level, `⏱️ ${description} completed in ${duration}s`);
    }
    
    // Context logging
    logContext(level, context, message) {
        this.log(level, `[${context}] ${message}`);
    }
}

// Factory function
function setupLogging(scriptName, options = {}) {
    return new TrilliumLogger(scriptName, options);
}

// Export for use in other modules
module.exports = {
    TrilliumLogger,
    setupLogging,
    SCRIPT_COLORS,
    LEVEL_COLORS,
    LOG_LEVELS
};

// CLI usage example
if (require.main === module) {
    const logger = setupLogging('test_script');
    
    logger.info('🚀 Testing JavaScript logging configuration');
    logger.debug('This is a debug message');
    logger.warn('This is a warning message');
    logger.error('This is an error message');
    logger.critical('This is a critical message');
    
    const startTime = Date.now();
    setTimeout(() => {
        logger.logExecutionTime('Test operation', startTime);
        logger.info('🏁 Logging test completed');
    }, 100);
}